Only build openvpn when changing source code or build files

mattock opened the pull request:
This PR adds a new Change Filter to the openvpn builds. The Change Filter passes a Change instance to a simple lambda function. The lambda function iterates over all the file list (c.files) and checks whether any of the entries matches a regular expression pattern. The regular expression lists all the filename patters that should trigger a build.

If none of the files in the Change do not match the regular expression then a build is not triggered by the Change. Otherwise a build is triggered.

Which files should trigger a build is up for discussion. I picked a selection that seemed reasonable at a quick glance.

References:
* https://docs.buildbot.net/current/manual/configuration/misc/change_filter.html
* https://github.com/buildbot/buildbot/blob/master/master/buildbot/changes/changes.py#L72

mattock commented:
Here's a simple standalone Python script for testing out the regular expression with various inputs without having to test with buildbot:

```
#!/usr/bin/env python3

import re

c = {}
c['files'] = ["README.md", "foo.txt", "file.cy", "src/tapctl/main.c", "src/openvpn/main.c"]

openvpn_filter_fn = r'^(src/(openvpn)/(\w|-)+\.(c|h)|m4/(\w|-)+\.m4|compat.m4|configure.ac|Makefile.am|version.m4)$'

filter_fn = lambda c: any([re.search(openvpn_filter_fn, f) for f in c['files']])

result = filter_fn(c)
print(result)  
```

flichtenheld commented on buildbot-host/buildmaster/master.cfg:
```diff
@@ -504,10 +505,15 @@ for factory_name, factory in factories.items():
 c['schedulers'] = []
+
+# Only build if any of the files in the Change match this regular expression
+openvpn_filter_fn = r'^(src/(openvpn)/(\w|-)+\.(c|h)|m4/(\w|-)+\.m4|compat.m4|configure.ac|Makefile.am|version.m4)$'
```
That regex definitely needs more work. Some initial ideas:

- Needs to cover all relevant Makefile.am files, not just top-level
- Needs to cover tests/
- Needs to cover src/compat/
- Needs to cover include/
- Needs to cover config.h.*

mattock replied:
Asking about this in tomorrow's community meeting would probably be quickest. I won't make it, but "what modifications should trigger a build" is a fairly straightforward question I think :smile:. So, I'll add it to the topic list.

flichtenheld commented on buildbot-host/buildmaster/master.cfg:
```diff
@@ -504,10 +505,15 @@ for factory_name, factory in factories.items():
                         change_filter=util.ChangeFilter(branch  = openvpn_branch,
-                                                        project = 'openvpn'),
+                                                        project = 'openvpn',
+                                                        filter_fn = lambda c: any([re.search(openvpn_filter_fn, f) for f in c.files])),
```
Honestly this scheduler does a build every few days currently. So not really worth implementing it here. The Gerrit scheduler would probably be the more interesting use case here.

mattock replied:
Ok. I can't test the Gerrit scheduler easily, but I think the filter can be moved over there without any modifications. The Change object should look the same anyways.

flichtenheld replied:
Just saw on https://docs.buildbot.net/latest/manual/configuration/changesources.html#gerriteventlogpoller that we probably need to add get_files argument to GerritEventLogPoller. Just noting it here to have it recorded it somewhere.

mattock commented:
I noticed that another option would be to pass the lambda function directory to the filesIsImportant parameter of the scheduler (see https://docs.buildbot.net/current/manual/configuration/schedulers.html).

The only difference is that if we use *fileIsImportant* then the skipped builds are shown on the status page. With the current Change Filter-based approach the skipped build is not shown in the status display. However, even with the Change Filter the Change itself is shown under *Builds/Last Changes*.

So, a matter of taste/style I guess.

flichtenheld commented:
> I noticed that another option would be to pass the lambda function directory to the filesIsImportant parameter of the scheduler (see https://docs.buildbot.net/current/manual/configuration/schedulers.html).
> 
> The only difference is that if we use _fileIsImportant_ then the skipped builds are shown on the status page. With the current Change Filter-based approach the skipped build is not shown in the status display. However, even with the Change Filter the Change itself is shown under _Builds/Last Changes_.
> 
> So, a matter of taste/style I guess.

I think I would prefer _fileIsImportant_, since that seems to exactly match what we're looking at

uddr commented:
With the Gerrit changes I would suggest to take a look at `event.patchSet.kind` (http://buildbot-host.openvpn.in:8010/#/builders/365/builds/693)

Gerrit patchsets have several [kinds of change uploaded](https://gerrit-documentation.storage.googleapis.com/Documentation/2.12.3/json.html#patchSet)
Sometimes there are patchsets which have only commit message update (NO_CODE_CHANGE) or no changes at all (NO_CHANGE). I think we can skip the builds for such a cases.

The only I'm not sure if there will be correct checks status displayed on Gerrit side (something to double check)

mattock commented:
I added a few changes:
* Pass the same filter_fn to Gerrit builds (untested, but "should work"(tm))
* Split the regular expression patterns into a list for readability
* Add missing files/directories to the patterns list
* Relaxed the regex patterns a bit (e.g. to catch Makefile.am changes)
    * This is mainly to keep them readable
    * The caveat is that every few years somebody might update a sample or README file in, say, "tests" subdirectory and trigger a build. For the record, last changes to those particular README files was in 2016.

I need to spin up a Gerrit instance so that I can experiment it without going around blindfolded. I think that's not required for this PR, though.

mattock commented:
I did not switch to *fileIsImportant* yet - nothing was decided in the [previous community meeting](https://www.mail-archive.com/openvpn-devel@lists.sourceforge.net/msg28127.html) and people may have differing opinions there. Let's bring it up in the community meeting tomorrow and try to reach a conclusion. Hopefully we can go through the "list of files that should trigger a build" as well.

flichtenheld commented on buildbot-host/buildmaster/master.cfg:
```diff
@@ -504,17 +505,34 @@ for factory_name, factory in factories.items():
+                         "config.h\..*",
+                         "configure.ac",
+                         "Makefile.am",
+                         "version.m4"]
```
Need to include `doc/.*\.rst` and `doc/Makefile.am` for the man-page/HTML builds.
Otherwise I'm fine with the list now.

mattock replied:
I added `doc/Makefile.am` and `.*\.rst`. While most .rst files are under *doc*, I wanted to cover *Changes.rst* and *CONTRIBUTING.rst* as well.

Looks good now?

mattock replied:
I'll resolve the merge conflict. Missed it the last time.

mattock replied:
Rebased against latest master and tested. Seems to work as intended. A DependentScheduler does not get triggered if the change_filter of their corresponding SingleBranchScheduler ignored as change. Which is what I'd expect.

mattock committed:
Buildbot: only run builds on relevant files

Signed-off-by: Samuli Seppänen <samuli.seppanen@gmail.com>

flichtenheld reviewed (approved):
LGTM. Did you get a chance to test the Gerrit part of the change or do we still need to do that?

mattock committed:
Buildmaster: get changed files from Gerrit

Without this change to Gerrit would never trigger a build because change filters
would never match.

Signed-off-by: Samuli Seppänen <samuli.seppanen@gmail.com>
